Cabal/src/Distribution/Backpack/UnifyM.hs

```diff
@@ -386,32 +386,39 @@ lookupMooEnv (m, i) k =
 
 -- The workhorse functions
 
-convertUnitIdU' :: MooEnv -> UnitIdU s -> UnifyM s OpenUnitId
+-- | Returns `OpenUnitId` if there is no a mutually recursive unit.
+-- | Otherwise returns a list of signatures instantiated by given `UnitIdU`.
+convertUnitIdU' :: MooEnv -> UnitIdU s -> UnifyM s (Either [ModuleName] OpenUnitId)
 convertUnitIdU' stk uid_u = do
     x <- liftST $ UnionFind.find uid_u
     case x of
-        UnitIdThunkU uid -> return (DefiniteUnitId uid)
+        UnitIdThunkU uid -> return $ Right (DefiniteUnitId uid)
         UnitIdU u cid insts_u ->
             case lookupMooEnv stk u of
-                Just _i ->
-                    failWith (text "Unsupported mutually recursive unit identifier")
-                    -- return (UnitIdVar i)
+                Just _i -> return $ Left $ Map.keys insts_u
                 Nothing -> do
                     insts <- for insts_u $ convertModuleU' (extendMooEnv stk u)
-                    return (IndefFullUnitId cid insts)
+                    return $ Right (IndefFullUnitId cid insts)
 
 convertModuleU' :: MooEnv -> ModuleU s -> UnifyM s OpenModule
 convertModuleU' stk mod_u = do
     mod <- liftST $ UnionFind.find mod_u
     case mod of
         ModuleVarU mod_name -> return (OpenModuleVar mod_name)
         ModuleU uid_u mod_name -> do
-            uid <- convertUnitIdU' stk uid_u
-            return (OpenModule uid mod_name)
+            x <- convertUnitIdU' stk uid_u
+            case x of
+                Right uid -> return (OpenModule uid mod_name)
+                Left mod_names ->
+                    let sigsList = hcat $ punctuate (text ", ") $ map (quotes . pretty) mod_names in
+                    failWith $
+                        text "Cannot instantiate requirement " <+> quotes (pretty mod_name) $$
+                        text "Check that \"build-depends:\" doesn't include library with signatures: " <+> sigsList $$
+                        text "as this creates a cyclic dependency, which GHC does not support."
```

Comment:
Could this and the previous occurrence be factored out? That would make it apparent in the code that a similar thing is outputted in two similar cases.